### DISCUSSION TITLE
multiple modification of the same file in parallel ?

### DISCUSSION THREAD

**Discussion by fredmo** [ImageMagick/ImageMagick#5915] [Help] [answered]
### ImageMagick version

 6.9.11-60 Q16 x86_64 2021-01-25

### Operating system, version and so on

WSL2

### Description

Hello,

I try to optimize a script where I modify my main image, but I don't find how to do it

By example : 
composite -geometry +$X+$Y -blend 50 A.png  C.png **C.png**
composite -geometry +$X2+$Y2 -blend 50 B.png  C.png **C.png**

By executing in background "&" or with the "parallel" command.
I have 
composite-im6.q16: missing an image filename `C.png' @ error/composite.c/CompositeImageCommand/1600.

I guess the first execution is doing a recreation of the C.png to compose the image and the C.png is temporarily erased.

Is there a way to do the command with only 2 filenames arguments?  "compose xxx fileA in fileC " ?  
because the current syntax seems to be " compose  fileA with fileC and store the result in fileC "

Regards,
Frederic

**Comment by snibgo** [comment]
You have two commands that write to the same file. This is fine if one command is executed after the other. But trying to execute the commands at the same time won't work. For example, while the first command is writing C.png, the second command cannot read that file, or write it.

> Is there a way to do the command with only 2 filenames arguments? "compose xxx fileA in fileC " ?

No. Even if there was, this wouldn't help you because IM would still need to do the same work: read both files, do some processing, and write the output.

> because the current syntax seems to be " compose fileA with fileC and store the result in fileC "

Correct.

For better performance, you could combine both commands into a single command.

**Comment by fredmo** [comment]
Thank you, I will try to combine commands.

Would you have an example ? 
I tried different combinations but it s still ko.
composite -geometry +960+64 A.png -geometry +960+44 B.png  C.png C.png 


Feel free to close this "Discussion"

**Comment by snibgo** [comment]
What version of IM do you use, on what platform? I will assume IM v7 on bash.

Instead of "composite", I suggest using "magick" with a "-composite" option. This needs the input images to be in the opposite order. Also, the blend argument is different (see the example in https://imagemagick.org/script/compose.php). So the commands would be:
```
magick C.png A.png -geometry +$X+$Y -compose Blend -define compose:args=50 -composite C.png
magick C.png B.png -geometry +$X2+$Y2 -compose Blend -define compose:args=50 -composite C.png
```
Now we can combine the two commands (untested):
```
magick \
  C.png A.png \
  -geometry +$X+$Y \
  -compose Blend -define compose:args=50 -composite \
  B.png \
  -geometry +$X2+$Y2 \
  -compose Blend -define compose:args=50 -composite \
  C.png
```

**Comment by fredmo** [comment, marked as answer]
Perfect with 

```
convert \
  C.png A.png \
  -geometry +$X+$Y \
  -compose Blend -define compose:args=50 -composite \
  B.png \
  -geometry +$X2+$Y2 \
  -compose Blend -define compose:args=50 -composite \
  C.png
```

I don't find the "magick" on the Ubuntu / WSL2.

**Reply by fmw42** [reply]
magick is the replacement for convert in IM 7.  Use magick with IM  7 and convert for IM 6.
